factom_keys/id.py

```diff
@@ -0,0 +1,197 @@
+import ed25519
+from bitcoin import base58
+from hashlib import sha256
+
+PREFIX_LENGTH = 3
+CHECKSUM_LENGTH = 4
+BODY_LENGTH = 35
+TOTAL_LENGTH = 39
+
+
+def generate_id_pairs():
+    """
+    Generate a set of Factom ID keypairs, sk1-sk4 and id1-id4.
+
+    :return: a list of tuples containing a random id key pair: (IDPrivateKey, IDPublicKey)
+    """
+    pairs = []
+
+    for i in range(0, 4):
+        signer, verifier = ed25519.create_keypair()
+        pair = (IDPrivateKey(signer.to_seed()), IDPublicKey(verifier.to_bytes()))
+        pairs.append(pair)
+
+    return pairs
+
+
+def _to_base58_string(prefixed_key: bytes):
+    """
+    Convert prefixed_key bytes into sk1/id strings with a checksum.
+
+    :param prefixed_key: the sk1 private key or id public key prefixed with the appropriate bytes
+    :return: a sk1 private key string or id public key string
+    """
+    prefix = prefixed_key[:PREFIX_LENGTH]
+    assert prefix in IDPublicKey.PREFIXES or prefix in IDPrivateKey.PREFIXES, "Invalid key prefix."
+    temp_hash = sha256(prefixed_key[:BODY_LENGTH]).digest()
+    checksum = sha256(temp_hash).digest()[:CHECKSUM_LENGTH]
+    return base58.encode(prefixed_key + checksum)
+
+
+class BadKeyStringError(Exception):
+    pass
+
+
+class IDPrivateKey:
+
+    # Prefixes for sk1 - sk4.
+    PREFIXES = [b"\x4d\xb6\xc9", b"\x4d\xb6\xe7", b"\x4d\xb7\x05", b"\x4d\xb7\x23"]
+
+    def __init__(self, seed_bytes=None, key_string=None):
+        assert (seed_bytes and not key_string) or (
+            not seed_bytes and key_string
+        ), "Only provide one of seed_bytes or key_string, not both"
+
+        if key_string:
+            if not IDPrivateKey.is_valid(key_string):
+                raise BadKeyStringError
+            decoded = base58.decode(key_string)
+            seed_bytes = decoded[PREFIX_LENGTH:BODY_LENGTH]
+
+        assert isinstance(seed_bytes, bytes)
+        assert len(seed_bytes) == 32
+        self.__signer = ed25519.SigningKey(seed_bytes)
+
+    @property
+    def key_bytes(self):
+        """
+        :return: the 32 byte raw private key
+        """
+        return self.__signer.to_seed()
+
+    def to_string(self, key_num: int):
+        """
+        :param key_num: the number of the ID key pair: x in (skx, idx) where 0 < x < 5
+        :return: the ID private key as a human-readable string in skx format
+        """
+        # Key number must be 1-4 for sk1-sk4
+        assert 0 < key_num < 5, "Key number must be 1-4 for keys sk1-sk4."
+
+        secret_body = IDPrivateKey.PREFIXES[key_num - 1] + self.key_bytes
+        return _to_base58_string(secret_body)
+
+    def get_public_key(self):
+        """
+        Derive and return the corresponding IDPublicKey
+
+        :return: the IDPublicKey corresponding to this IDPrivateKey
+        """
+        public_bytes = self.__signer.get_verifying_key().to_bytes()
+        return IDPublicKey(public_bytes)
```

Comment:
Let's have this one first create the `ServerIdentityPublicKey`, then call `set_key_level()` for the public key using the same key level as found on this private key